src/Orleans/Threading/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Concurrency;
+using Orleans.Runtime;
+
+namespace Orleans.Threading
+{
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private readonly Dictionary<Guid, CancellationTokenSource> _cancellationTokenSources;
```

Comment:
I've tested the ovehead of keeping cts's in memory with using `WeakReference` and the difference was 20%. So I decided that using it not worth the code coplication, as Guid's will still remain in mmomory without compaction method. As for `ConditionalWeakTable` - it checks references to the keys, in our case it's guids.